File: scala/robot-simulator/src/main/scala/Robot.scala
object Bearing extends Enumeration {
  type Bearing = Value
  val North, East, South, West = Value

  def nextRight(bearing: Bearing) =
    bearing match {
      case North => East
      case East => South
      case South => West
      case West => North
    }

  def nextLeft(bearing: Bearing) =
    bearing match {
      case North => West
      case East => North
      case South => East
      case West => South
    }
}

import Bearing._

case class Robot(bearing: Bearing, coordinates: (Int, Int)) {
  def turnLeft() = Robot(nextLeft(bearing), coordinates)
  def turnRight() = Robot(nextRight(bearing), coordinates)
  def advance = copy(coordinates = advanceCoordinates)
  def advanceCoordinates =
    bearing match {
      case North => (coordinates._1, coordinates._2 + 1)
      case East => (coordinates._1 + 1, coordinates._2)
      case South => (coordinates._1, coordinates._2 - 1)
      case West => (coordinates._1 - 1, coordinates._2)
    }

  def simulate(input: String) =
    input.foldLeft(this){
      case (robot, 'A') => robot.advance
      case (robot, 'L') => robot.turnLeft
      case (robot, 'R') => robot.turnRight
    }
}

File: haskell/robot-simulator/src/Robot.hs
module Robot
    ( Bearing(East,North,South,West)
    , bearing
    , coordinates
    , mkRobot
    , move
    ) where

prev :: (Eq a, Enum a, Bounded a) => a -> a
prev x
  | x == minBound = maxBound
  | otherwise = pred x

next :: (Eq a, Enum a, Bounded a) => a -> a
next x
  | x == maxBound = minBound
  | otherwise = succ x

data Bearing = North
             | East
             | South
             | West
             deriving (Eq, Show, Enum, Bounded)

data Robot = Robot { bearing :: Bearing
                   , coordinates :: (Integer, Integer)
                   } deriving (Eq, Show)

mkRobot :: Bearing -> (Integer, Integer) -> Robot
mkRobot = Robot

turnLeft :: Bearing -> Bearing
turnLeft = prev

turnRight :: Bearing -> Bearing
turnRight = next

move :: Robot -> String -> Robot
move = foldl (flip moveOn)
  where
    advance North x y = (x, y+1)
    advance East x y = (x+1, y)
    advance South x y = (x, y-1)
    advance West x y = (x-1, y)

    turn robot f = robot { bearing = f $ bearing robot }

    move' robot f = robot { coordinates = f $ coordinates robot }

    moveOn 'L' robot = robot `turn` turnLeft
    moveOn 'R' robot = robot `turn` turnRight
    moveOn 'A' robot = robot `move'` uncurry (advance $ bearing robot)
    moveOn _ _ = error "unexpected case"
